driver-core/src/main/com/mongodb/client/model/expressions/MqlExpression.java

```diff
@@ -382,22 +380,20 @@ public <R extends Expression> ArrayExpression<R> isArrayOr(final ArrayExpression
         return (ArrayExpression<R>) this.isArray().cond(this.assertImplementsAllExpressions(), other);
     }
 
-    public BooleanExpression isDocument() {
+    public BooleanExpression isDocumentOrMap() {
         return new MqlExpression<>(ast("$type")).eq(of("object"));
     }
 
     @Override
     public <R extends DocumentExpression> R isDocumentOr(final R other) {
-        return this.isDocument().cond(this.assertImplementsAllExpressions(), other);
-    }
-
-    public BooleanExpression isMap() {
-        return new MqlExpression<>(ast("$type")).eq(of("object"));
+        return this.isDocumentOrMap().cond(this.assertImplementsAllExpressions(), other);
     }
 
+    @SuppressWarnings("unchecked")
     @Override
-    public <R extends Expression> MapExpression<R> isMapOr(final MapExpression<R> other) {
-        return this.isMap().cond(this.assertImplementsAllExpressions(), other);
+    public <R extends Expression> MapExpression<R> isMapOr(final MapExpression<? extends R> other) {
+        MqlExpression<Expression> isMap = (MqlExpression<Expression>) this.isDocumentOrMap();
+        return newMqlExpression(isMap.ast("$cond", this.assertImplementsAllExpressions(), other));
     }
```

Comment:
`this.isDocumentOrMap()` produces `BooleanExpression`. When we cast it to its implementation, the type argument is inapplicable, and so it does not make sense as much sense to cast to `MqlExpression<Expression>` as to cast to `MqlExpression<IntegerExpression>` - both are meaningless. We should cast to `MqlExpression<?>` as not only it makes sense, but also this cash is checked and we don't need `@SuppressWarnings("unchecked")` anymore:


```java
@Override
public <R extends Expression> MapExpression<R> isMapOr(final MapExpression<? extends R> other) {
    MqlExpression<?> isMap = (MqlExpression<?>) this.isDocumentOrMap();
    return newMqlExpression(isMap.ast("$cond", this.assertImplementsAllExpressions(), other));
}
```

Reply:
The method `isMapOr ` is still marked with `@SuppressWarnings("unchecked")`.